jme3-xr/src/main/java/com/jme3/input/xr/Eye.java

```diff
@@ -0,0 +1,94 @@
+import com.jme3.math.Vector3f;
+import com.jme3.renderer.Camera;
+import com.jme3.renderer.ViewPort;
+import com.jme3.scene.Geometry;
+import com.jme3.scene.shape.Box;
+import com.jme3.texture.FrameBuffer;
+import com.jme3.texture.Texture;
+import com.jme3.texture.Texture2D;
+import com.jme3.texture.FrameBuffer.FrameBufferTarget;
+import com.jme3.texture.Image.Format;
+
+public class Eye {
+	static int index = 0;
+	SimpleApplication app;
+	float posX;
+	Vector3f tmpVec = new Vector3f();
+	Texture2D offTex;
+	Geometry offGeo;
+	Camera offCamera;
+	Vector3f centerPos = new Vector3f(0f, 0f, -5f);
+	Quaternion centerRot = new Quaternion();
+    
+    public Eye(SimpleApplication app)
+    {
+    	this.app = app;
+    	setupOffscreenView(app);
+        Material mat = new Material(app.getAssetManager(), "Common/MatDefs/Misc/Unshaded.j3md");
+        mat.setTexture("ColorMap", offTex);
+        
+        offGeo = new Geometry("box", new Box(1, 1, 1));
+        offGeo.setMaterial(mat);
+    }
+    
+    public void setPosX(float posX) { this.posX = posX; }
+    public float getPosX() { return posX; }
+    
+    /** Moves the camera.
+     * @param The new absolute center position. */
+    public void moveAbs(Vector3f newPos)
+    {
+    	centerPos.set(newPos);
+    	rotateAbs(offCamera.getRotation());
+    }
+    
+    /** Rotates the camera, and moves left/right.
+     * @param The new rotation. */
+    public void rotateAbs(Quaternion newRot)
+    {
+    	tmpVec.set(posX, 0.0f, 0.0f);
+    	newRot.multLocal(tmpVec);
+    	offCamera.setLocation(tmpVec.addLocal(centerPos));
+    	offCamera.setRotation(newRot);
+    }
+    
+    private void setupOffscreenView(SimpleApplication app)
+    {
+    	int w = app.getContext().getSettings().getWidth();
+    	int h = app.getContext().getSettings().getHeight();
+        offCamera = new Camera(w, h);
+
+        ViewPort offView = app.getRenderManager().createPreView("OffscreenViewX" + (index++), offCamera);
+        offView.setClearFlags(true, true, true);
+        offView.setBackgroundColor(ColorRGBA.DarkGray);
+        FrameBuffer offBuffer = new FrameBuffer(w, h, 1);
+
+        //setup framebuffer's cam
+        offCamera.setFrustumPerspective(45f, 1f, 1f, 1000f);
+        offCamera.lookAt(new Vector3f(0f, 0f, 0f), Vector3f.UNIT_Y);
+
+        //setup framebuffer's texture
+        offTex = new Texture2D(w, h, Format.RGBA8);
+        offTex.setMinFilter(Texture.MinFilter.Trilinear);
+        offTex.setMagFilter(Texture.MagFilter.Bilinear);
+
+        //setup framebuffer to use texture
+        offBuffer.setDepthTarget(FrameBufferTarget.newTarget(Format.Depth));
+        offBuffer.addColorTarget(FrameBufferTarget.newTarget(offTex));
+
+        //set viewport to render to offscreen framebuffer
+        offView.setOutputFrameBuffer(offBuffer);
+        offView.attachScene(app.getRootNode());
+    }
+    
+    public void render()
+    {
+    	app.getRenderManager().renderGeometry(offGeo);
```

Comment:
What is this offgeo for exactly?

Reply:
Rendering offscreen onto texture of offgeo, and then render this offgeo on framebuffer of openxr was the only way I found, to get a 3d picture onto the HMD.
If we find an other way, we can change this of course.
But it renders very smooth.

Reply:
You're right. I tried removing it and it stopped working. I wonder if rendering the geometry triggers all the dependant materials and textures to be rendered and that's what actually needed?

I'm going to have a play and see if that can be streamlined